Make SSL mode configurable for internal and external databases

zackkrida opened the pull request:
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Related to #4847

<!-- Concisely describe what the pull request does. -->

**Technical details**
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->

This PR makes the ssl mode for Postgres connections configurable via environment variable (for internal db connections) and via the UI (for connecting to existing data databases). 

This also allows the ssl mode to be edited for existing DBs.

This PR does **not** support configuring the `PGSSLROOTCERT` or the various `verify-` ssl modes. That's a more advanced setting that adds complexity for typical users, so we should only add it based on user requests for the missing functionality. 

**Screenshots**
<!-- Add screenshots to show the problem and the solution; or delete the section entirely. -->

<img width="1903" height="1082" alt="Screenshot From 2025-12-02 15-14-25" src="https://github.com/user-attachments/assets/75e4e957-ac81-4c74-a78c-747aec67cf62" />
<img width="1047" height="843" alt="Screenshot From 2025-12-02 15-14-09" src="https://github.com/user-attachments/assets/1a6df9e2-b4a5-455a-8d5a-fd259c502ee9" />

Preview of new docs:

<img width="1077" height="630" alt="image" src="https://github.com/user-attachments/assets/b0c9d4e5-1245-4ad4-a436-e7c6658e7dc7" />

<img width="1051" height="823" alt="image" src="https://github.com/user-attachments/assets/f7d35554-3446-4a1c-a006-81fd25a46d66" />

## Testing Instructions

To make this easier to test, I set up a gist for running a DB with SSL required locally. You should be able to try to connect to this in the Mathesar UI and see an error when you set SSL mode to disabled:

https://gist.github.com/zackkrida/dda14152cb5182982542085690d64373

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

zackkrida committed:
Make SSL mode configurable for internal and external databases

zackkrida committed:
Add sslmode to edit db form

zackkrida commented:
@mathemancer here's an alternative to #4981 that I think will be better for users as it allows them to set the ssl mode per server and edit the ssl mode, rather than hardcode a single ssl mode for all connections. 

@seancolsen, the UI changes here are pretty minimal but please review them nonetheless.

zackkrida committed:
Add sslmode to db setup mock

mathemancer commented on mathesar/utils/permissions.py:
```diff
@@ -34,21 +34,25 @@ def set_up_new_database_for_user_on_internal_server(
             "Mathesar can't be installed in the internal database."
         )
+    # Use sslmode from internal database config if available
+    sslmode = getattr(conn_info, 'sslmode', None) or 'prefer'
```
I'd prefer `try/except` here. Also, did you run into a case where `sslmode` _isn't_ an attribute of `conn_info`? I don't see how that's possible. Finally, why not just use `'prefer'` as the default?

mathemancer commented on mathesar/utils/permissions.py:
```diff
@@ -62,7 +66,8 @@ def set_up_new_database_for_user_on_internal_server(
 @transaction.atomic
 def set_up_preexisting_database_for_user(
-        host, port, database_name, nickname, role_name, password, user, sample_data=[]
+        host, port, database_name, nickname, role_name, password, user,
```
We typically aren't wrapping in this style in this project. Either one line for the whole args list, or separate lines for each name (my preference), please.

mathemancer commented on mathesar/utils/permissions.py:
```diff
@@ -87,9 +93,17 @@ def set_up_preexisting_database_for_user(
+        host=host, port=port,
+        defaults={"sslmode": sslmode}
+    )
+    # Update sslmode if the server already exists but sslmode differs
```
Remove this comment, please.

mathemancer commented on mathesar/utils/permissions.py:
```diff
@@ -34,21 +34,25 @@ def set_up_new_database_for_user_on_internal_server(
         raise BadInstallationTarget(
             "Mathesar can't be installed in the internal database."
         )
+    # Use sslmode from internal database config if available
```
Remove this comment please.

mathemancer commented on mathesar/utils/permissions.py:
```diff
@@ -87,9 +93,17 @@ def set_up_preexisting_database_for_user(
+        sslmode='prefer'
 ):
-    server, _ = Server.objects.get_or_create(host=host, port=port)
+    server, _ = Server.objects.get_or_create(
```
Couldn't this just be an `update_or_create` ? The pattern below is a bit unwieldy.

mathemancer reviewed (changes requested):
I made some specific comments. Mostly around style, but I'd much rather use the `update_or_create` function where I noted.

Also, there's a lot of comments explaining simple logic. Please remove those to reduce clutter.

zackkrida committed:
Address review feedback from Brent

zackkrida commented:
Thanks for the review @mathemancer. 

I've switched to `update_or_create`, removed all the redundant comments--a habit I abhor from others but still do myself at times when writing python--, fixed the formatting, and made 'prefer' the default in `PostgresConfig` so we can just access `conn_info.sslmode` directly.

zackkrida committed:
Merge branch 'develop' into ssl

zackkrida committed:
Fix API tests

seancolsen committed:
Refactor SSL mode options

seancolsen committed:
Link to user guide from in-app help text

seancolsen reviewed (approved):
Docs changes and front end changes look good!

- I pushed df2167b7b0e515f0580b434a94a64bee2395df2d to address a small user-facing issue.

    I like the markdown documentation you wrote about this new setting! So I changed the in-app help text to link out to the user guide. I think this means we don't need quite as much content in the help text because readers can get that in the guide by following the link if they need it.

    Before:

    <img width="600" src="https://github.com/user-attachments/assets/84937e47-fd2a-4a0e-b782-bf882e165c58" />

    After:

    <img width="600" src="https://github.com/user-attachments/assets/f8432154-cd49-49b4-840b-4763525bcd22" />


- I pushed 300c4e894e047d6c984bf1758e69e5ffe57dacc5 to address a small developer-facing issue.

    In `SelectSslMode.svelte` you had:

    ```ts
    $_(option.labelKey)
    ```

    We've been actively avoiding using dynamic i18n keys like this. And (to my knowledge) all i18n function arguments across the codebase are currently static strings. Why? We have some tooling that assumes static strings. If you look at `check_unused_strings.py` you'll see that we're searching the codebase for static strings inside translation functions. This helps us identify orphaned strings in `en/dict.json`.

    This is essentially an undocumented code standard. I really wanted us to have an i18n developer guide immediately after implementing the i18n system, but we've not chosen to prioritize work on that. Oh well.

    While fixing this code standard, I also refactored the way the options were defined. The code is now a bit simpler.

I didn't review backend changes.

zackkrida commented:
Those changes look great; thanks @seancolsen!

zackkrida committed:
Merge branch 'develop' into ssl

mathemancer reviewed (approved):
Okay, looks good to go from my end.
